File: 241.为运算表达式设计优先级.py
#
# @lc app=leetcode.cn id=241 lang=python3
#
# [241] 为运算表达式设计优先级
#

# @lc code=start
class Solution:
    def diffWaysToCompute(self, input: str) -> List[int]:
        if input.isdigit():
            return [int(input)]

        result = []
        for i, c in enumerate(input):
            if c in ['+', '-', '*']:
                left = self.diffWaysToCompute(input[:i])
                right = self.diffWaysToCompute(input[i+1:])
                for l in left:
                    for r in right:
                        if c == '+':
                            result.append(l+r)
                        if c == '-':
                            result.append(l-r)
                        if c == '*':
                            result.append(l*r)
        return result

# @lc code=end

File: 241.为运算表达式设计优先级.cpp
/*
 * @lc app=leetcode.cn id=241 lang=cpp
 *
 * [241] 为运算表达式设计优先级
 */

// @lc code=start
class Solution
{
public:
    vector<int> diffWaysToCompute(string input)
    {
        int num = 0;
        bool flag = true;
        vector<int> result;
        for (char c : input)
        {
            flag = isdigit(c);
            if (flag)
                num = num * 10 + (c - '0');
            else
                break;
        }
        if (flag)
        {
            result.push_back(num);
            return result;
        }
        for (int i = 0; i < input.size(); i++)
        {
            if ('+' == input[i] || '-' == input[i] || '*' == input[i])
            {
                vector<int> left = diffWaysToCompute(input.substr(0, i));
                vector<int> right = diffWaysToCompute(input.substr(i + 1, input.size()));
                for (int l : left)
                {
                    for (int r : right)
                    {
                        if ('+' == input[i])
                            result.push_back(l + r);
                        if ('-' == input[i])
                            result.push_back(l - r);
                        if ('*' == input[i])
                            result.push_back(l * r);
                    }
                }
            }
        }
        return result;
    }
};
// @lc code=end
